### ISSUE TITLE
Allow to scan offline (not open the products)

### ISSUE THREAD

**Issue by teolemon** [openfoodfacts/openfoodfacts-androidapp#2069]
I've created a special excerpt of the database, which has the required data to scan products for France. This is 11 Mb zipped, and 37Mb unzipped.
http://fr.openfoodfacts.org/data/offline/fr.openfoodfacts.org.products.small.csv.tgz
Can we download this data and store it in the DB ?

**Comment by teolemon** [comment]
I can create it for other countries on demand.

**Comment by prashantkh19** [comment]
We can download the file on the first run and will store it in the DB. I will try that.

**Comment by prashantkh19** [comment]
@teolemon please review... as of now, since the data is too big (~700000 products), so it's taking too much time to store in the database.

**Comment by teolemon** [comment]
@jinqian @huzaifaiftikhar @PrajwalM2212 any ideas to massively increase it client side ?
Can we perhaps generate and store a separate db server side ?

**Comment by teolemon** [comment]
@prashantkh19 Did you manage to load it completely ? if so, how long did it take ?

**Comment by prashantkh19** [comment]
No, I didn't. I did for a subset of 10,000 products and it took around 5-6 minutes.

**Comment by teolemon** [comment]
Do you have an idea of the size of the DB for 10k products ?

**Comment by PrajwalM2212** [comment]
Your code is trying to download it during the first run. That is great but I am more worried about the time that will take.  
Even at present without the database the splash screen takes a lot of time to load.  In fact we must try to reduce the splash screen time as it is. 
I think we should allow the user to schedule a time for download if he wants offline capability just like how we can schedule software updates on the phone. This way we don't end up consuming more time or memory.

**Comment by PrajwalM2212** [comment]
What if we don't save the data in a database. When a user scans a product, we will 
Scan the csv file if the product exists and then fetch that record and display it to them. 
We can improve the organization of the csv file so that we take less time to fetch the product from csv 
@teolemon @prashantkh19

**Comment by PrajwalM2212** [comment]
@prashantkh19 That is why I was saying that we download like 10000 products at every attempt rather than as whole

**Comment by prashantkh19** [comment]
@PrajwalM2212 yeah, it can be good, we can divide the csv file based on sorted barcodes, each file having let say, 10000 products and a separate file which contains the metadata of all the csv files. That way, we will need to scan just 2 files for each product scan, and it will take hardly 5 seconds to get the results. And maybe we can try hashing each csv also.

**Comment by teolemon** [comment]
weird: http://greenrobot.org/wordpress/wp-content/uploads/2016-07-08-android-orm-performance-benchmark.png

**Comment by teolemon** [comment]
15k entities per second.

**Comment by prashantkh19** [comment]
Optimized Version: took around 4 minutes to `read csv`(~ 2 minutes) and `saving to the database`(~ remaining 2 minutes) apart from downloading time for ~700000 products (complete csv file data).

**Comment by prashantkh19** [comment]
It can still be optimized more, as I am saving a list of max 15000 entities each time as of now. If I increase the list size, then time can be reduced but increasing the size of the list may result in crashes due to memory. A better Data Structure (space optimized) than ArrayList may help or if we somehow can store `List<String[]> records` returned by `reader.readAll()` directly to greenDao database.

**Comment by prashantkh19** [comment]
@teolemon @PrajwalM2212 can you review the latest commit?

**Comment by serpico** [comment]
> I've created a special excerpt of the database, which has the required data to scan products for France. This is 11 Mb zipped, and 37Mb unzipped.
> http://fr.openfoodfacts.org/data/offline/fr.openfoodfacts.org.products.small.csv.tgz
> Can we download this data and store it in the DB ?

@teolemon what about optionally limiting the loading/excerpt to a specific store? Through the preferences maybe. I know I do most of my grocery shopping in one store so...
Also, is there a way to test your file directly from the phone with the current app ?

**Comment by teolemon** [comment]
@serpico That will be V2 or 3. For now, we can afford to load everything. In the future, we'll try to let people control how much detail they want (like full products) and what countries, what stores they want. Besides, right now, we don't know if a product is sold everywhere.
@prashantkh19 hasn't implemented scanning yet. All it does is load the DB right now :-)

**Comment by teolemon** [comment]
I have fixed the CSV export for NOVA which wasn't exported correctly:
https://world.openfoodfacts.org/data/offline/en.openfoodfacts.org.products.small.csv.zip

**Comment by teolemon** [comment]
I'm bumping this one to P1

**Comment by teolemon** [comment]
https://storage.googleapis.com/icons.google.com.a.appspot.com/google-assets/system_icons/quantum/file/res-export/ic_cloud_off_24px.svg

**Comment by kartikaysharma01** [comment]
@teolemon  Can you break this down in small tasks ?

**Comment by naivekook** [comment]
Hi!
I'm a new here, excuse me if I ask fool questions and tell obvious things and of course for hoping into the train in last moment :)

@teolemon Do you consider an approach with downloading the whole sqlite DB file?
I had a project where we chose this approach and quick googling shows that converting csv to sqlite on server side is quite easy task. The main idea is to download .sqlite database from serer as is, put it on sdcard and just work with it, without parsing csv file and copying all data to local DB.

From top of my head I see offline flow V1 next:

- [ ]  we have a screen with available regions to download
- [ ]  the app downloads the database and store somewhere info about region and DB version
- [ ]  once per N day the app check on server is there a new version of DB for user
- [ ]  we have toogle button in left menu "Enable offline mode" which tuns on/off search from local DB or from server

**Comment by kartikaysharma01** [comment]
I'd suggest we keep a ```Go Offline``` tab in the sidebar. On the landing page, we can keep a link to the tab , for better user awareness of the feature.
In the tab, we can give user a choice to which country's database he wants to download. And once the download button is clicked, we can run the process in the background with a notification.
I like the idea given by @naivekook , but I don't think parsing and copying data to local DB is a blocker issue as it is a one time process, and we can use server side sqlite database as and when it is available.

**Comment by VaiTon** [comment]
Also, we keep the size problem. I think the best approach would just to store a country into the local db with just a few attributes: name, ecoscore, nutriscore, brand and quantity maybe

**Comment by kartikaysharma01** [comment]
@VaiTon I think some API changes are required, Should I wait for that or continue with the API at hand ?
https://github.com/openfoodfacts/openfoodfacts-server/pull/4347

**Comment by kartikaysharma01** [comment]
@VaiTon  @teolemon  I would like to implement this. 

These are the subtasks :

- [ ] The user will have a option in preference to download the data for their country. 
- [ ] Parse and store the downloaded file in background and show user a progress bar.
- [ ] When the user scans a barcode in offline mode the data will be shown in bottom sheet but the user will unable to swipe it up.

For this we will need to store the nutriscore, nova group (or CO2) , eco score and ingredient analysis. 
Have I got this right?

**Comment by kartikaysharma01** [comment]
Also , can you please explain the CSV design.  @teolemon 
https://world.openfoodfacts.org/data/offline/en.openfoodfacts.org.products.small.csv.zip

**Comment by teolemon** [comment]
@kartikaysharma01 It's basically a very minified version of the database. 
code	product_name	quantity	brands	nutriscore_grade	nova_group	ecoscore_grade
8938506580127	Dien Bao sweet potato vermiceilli					
8938506854082	Virgin Coconut Oil	1 l	Vietcoco			
8938506854471	Lait De Coco Sterilisé				4.0	
8938506854518	Thanh vinh		Organic			
8938506854563	coconut cream	13.5 fl.oz	vietcoco		4.0

**Comment by teolemon** [comment]
And I believe the current export is broken, here's the Python script that converts the daily CSV export: https://github.com/openfoodfacts/openfoodfacts-server/blob/main/scripts/generate_dump_for_offline_apps_off.py
